File: frontend/src/types/finance.ts
export interface RevenueDataPoint {
  date: string;
  revenue: number;
  costs: number;
  profit: number;
}

export interface TokenUsageSummary {
  provider: string;
  model: string;
  total_tokens: number;
  prompt_tokens: number;
  completion_tokens: number;
  cost_usd: number;
  percent_of_total: number;
  request_count: number;
  period: string;
}

export interface VentureFinancials {
  venture: string;
  revenueThisMonth: number;
  revenueLastMonth: number;
  costsThisMonth: number;
  profitThisMonth: number;
  growthPercent: number;
  revenueYtd: number;
}

export interface MonthlySummary {
  month: string;
  revenue_aud: number;
  cost_usd: number;
  cost_aud: number;
  profit_aud: number;
  sales_count: number;
  tokens_used: number;
}

export interface CostForecast {
  month: string;
  actual: number | null;
  projected: number;
  lowerBound: number;
  upperBound: number;
}

export interface FinanceDashboard {
  period_start: string;
  period_end: string;
  total_revenue_aud: number;
  total_cost_usd: number;
  total_profit_aud: number;
  active_products: number;
  total_sales: number;
  avg_order_value_aud: number;
  token_usage: TokenUsageSummary[];
  revenue_by_platform: Array<{ platform: string; revenue_aud: number; sales_count: number; period: string }>;
  monthly_trend: MonthlySummary[];
  revenue_history: RevenueDataPoint[];
  model_costs: TokenUsageSummary[];
}

/** Legacy camelCase alias — kept for components not yet migrated */
export interface ModelCostBreakdown {
  model: string;
  provider: string;
  tokensUsed: number;
  costUsd: number;
  percentOfTotal: number;
  requestCount: number;
}

export function tokenUsageToModelCost(t: TokenUsageSummary): ModelCostBreakdown {
  return {
    model: t.model,
    provider: t.provider,
    tokensUsed: t.total_tokens,
    costUsd: t.cost_usd,
    percentOfTotal: t.percent_of_total,
    requestCount: t.request_count,
  };
}

export interface FinanceSummary {
  revenueThisMonthAud: number;
  revenueLastMonthAud: number;
  revenueGrowthPercent: number;
  costsThisMonthUsd: number;
  costsLastMonthUsd: number;
  costsGrowthPercent: number;
  profitThisMonthAud: number;
  tokenUsageTodayM: number;
  tokenUsageMonthM: number;
  monthlyTarget: number;
  targetProgressPercent: number;
}

File: backend/app/schemas/finance.py
"""Pydantic schemas for Finance endpoints."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class RevenueByPlatform(BaseModel):
    platform: str
    revenue_aud: float
    sales_count: int
    period: str


class MonthlySummary(BaseModel):
    month: str  # YYYY-MM
    revenue_aud: float
    cost_usd: float
    cost_aud: float  # Converted for display
    profit_aud: float
    sales_count: int
    tokens_used: int


class TokenUsageSummary(BaseModel):
    provider: str
    model: str
    total_tokens: int
    prompt_tokens: int
    completion_tokens: int
    cost_usd: float
    period: str


class ForecastPoint(BaseModel):
    month: str
    predicted_revenue_aud: float
    lower_bound: float
    upper_bound: float
    confidence: float


class RevenueForecast(BaseModel):
    horizon_months: int
    forecast: list[ForecastPoint]
    model_used: str
    generated_at: datetime


class FinanceDashboard(BaseModel):
    period_start: date
    period_end: date
    total_revenue_aud: float
    total_cost_usd: float
    total_profit_aud: float
    active_products: int
    total_sales: int
    avg_order_value_aud: float
    token_usage: list[TokenUsageSummary]
    revenue_by_platform: list[RevenueByPlatform]
    monthly_trend: list[MonthlySummary]


class CostRecordRead(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID | None
    task_id: uuid.UUID | None
    provider: str
    model: str
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    cost_usd: float
    trace_id: str | None
    label: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
